feat: add Service Worker interception

rwoll opened the pull request:
Adds Service Worker Main Script Interception (CR-only) and Service Worker related utils:

- (all browsers) `Response.fulfilledByServiceWorker`
- (all browsers) `serviceWorkers: "allow" | "block"` context option (and `test.use` option)
- (CR-only) instruments Service Worker Networking (allows for interception/routing, event listening, offline mode, etc.)

Resolves #1090
Resolves #14666
Supercedes #12565

Landed upstream patches:
- https://chromium-review.googlesource.com/c/chromium/src/+/3510917
- https://chromium-review.googlesource.com/c/chromium/src/+/3526571
- https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/3566669
- https://chromium-review.googlesource.com/c/chromium/src/+/3544685
- https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/3610924

Pending upstream patches (to resolve the failing test cases in this PR):
- https://chromium-review.googlesource.com/c/chromium/src/+/3689949

Since this Playwright change opts all users in by default, but the upstream pending patch is required to fix some rare failure modes, this Playwright PR should be followed up with a minor change that guards the installation of the new instrumentation based on CR major revision.

rwoll committed:
feat: service worker plumbing

rwoll commented on packages/playwright-core/src/server/chromium/crServiceWorker.ts:
```diff
@@ -0,0 +1,74 @@
+    });
+
+    // These might fail if the target is closed before we receive all execution contexts.
+    this._networkManager.initialize().catch(() => {});
```
We'll need to think about how to opt people in/out. Doing this unconditionally would risk old code triggering an error by calling `request.frame()`.

rwoll commented on packages/playwright-core/src/client/network.ts:
```diff
@@ -172,13 +173,25 @@ export class Request extends ChannelOwner<channels.RequestChannel> implements ap
   frame(): Frame {
+    if (!this._initializer.frame) {
+      assert(this.isServiceWorkerRequest());
+      throw new Error('Service Worker requests do not have an associated frame.');
```
Frame had been nullable, but then we made it non-nullable at some point in the past. We can't go back, and SW's don't have a frame associated with them.

Meir017 commented:
Intercepting only requests sent from the service worker is not possible through the first condition argument, I was able to do something like this:
```js
test('should intercept only serviceworker request, not page', async ({ context, page, }) => {
  await context.route('**/sw.js', route =>
    route.fulfill({
      contentType: 'text/javascript',
      status: 200,
      body: `
        self.addEventListener('fetch', event => {
          if (event.request.url.endsWith('/data.json')) {
            event.respondWith(fetch(event.request));
          }
        });
      `,
    })
  );
  await context.route('**/data.json', async route => {
    if (route.request().isServiceWorkerRequest()) {
      return route.fulfill({
        contentType: 'text/plain',
        status: 200,
        body: 'from sw',
      });
    } else {
      return route.continue();
    }
  })

  const [ sw ] = await Promise.all([
    context.waitForEvent('serviceworker'),
    page.goto(server.PREFIX + '/serviceworkers/empty/sw.html'),
  ]);
  const response = await page.evaluate(() => fetch('/data.json').then(r => r.text()));
  expect(response).toBe('from sw');
});
```

rwoll committed:
test.fail does not work with timedout tests

rwoll committed:
add SW fetch interception test

rwoll commented:
@Meir017 See this newly added test: https://github.com/microsoft/playwright/pull/14321/commits/92e28e0e7269597d58e4062f2329050b3a8b575f#diff-f48a7787820cb49396256c1234745152eb05ca5a3f43b6d128ce0ee45eb6dff6R253

It aims to cover the intent of your scenario and passes. Notably it: (1) waits for the SW to be active, and (2) asserts the SW actually intercepted the request (according to the SW). Let me know if this covers your case!

tl;dr: I think https://github.com/microsoft/playwright/pull/14321#issuecomment-1133755994 was invalid since it didn't wait for the SW to be active, so SW wasn't yet handling your request (which was still a page req at the time and was reported as such in the context.route).

rwoll commented on tests/library/chromium/chromium.spec.ts:
```diff
@@ -30,6 +30,233 @@ test('should create a worker from a service worker', async ({ page, server }) =>
+    const [inner] = await Promise.all([
+      context.waitForEvent('request', r => r.url().endsWith('/inner.txt')),
+      page.evaluate(() => fetch('/inner.txt')),
+    ]);
+    // FIXME: This doesn't seem to be the correct assetion.
+    // Ensure the reques is actually intercepted by SW.
+    expect(inner.isServiceWorkerRequest()).toBe(false);
+    expect(inner.serviceWorker()).toBe(null);
```
There's a bug somewhere! This is wrong. It should be considered service worker. Interestingly, if you add `context.route('**', (route) => route.continue());` to the top of the test, it passes. (Having the no-op context route does enable more CDP events.)

rwoll replied:
For SW with `fetch` listener, it looks like we are firing two requestWillBeSentEvents. One on Page and one on SW.

rwoll commented:
The following test has different behavior when the `context.route` line is commented vs. uncommented:

```ts
test.only('isServiceWorker() and serviceWorker() work', async ({ context, page, server }) => {
    // context.route('**', (route) => route.continue());
    const [worker] = await Promise.all([
      context.waitForEvent('serviceworker'),
      page.goto(server.PREFIX + '/serviceworkers/fetch/sw.html')
    ]);
    await page.evaluate(() => window['activationPromise']);
    const [swReq, frameReq] = await Promise.all([
      context.waitForEvent('request', r => r.url().endsWith('/inner.txt') && r.isServiceWorkerRequest()),
      context.waitForEvent('request', r => r.url().endsWith('/inner.txt') && !r.isServiceWorkerRequest()),
      page.evaluate(() => fetch('/inner.txt')),
    ]);
    const [ url ] = await worker.evaluate(() => self['intercepted']);
    expect(url).toMatch(/\/inner\.txt$/);
    expect(swReq.frame).toThrow();
    expect(swReq.isServiceWorkerRequest()).toBe(true);
    expect(swReq.serviceWorker()).toBe(worker);

    expect(frameReq.frame()).toBe(page.mainFrame());
    expect(frameReq.isServiceWorkerRequest()).toBe(false);
    expect(frameReq.serviceWorker()).toBe(null);
});
```

With the following log line:

```patch
diff --git a/packages/playwright-core/src/server/chromium/crNetworkManager.ts b/packages/playwright-core/src/server/chromium/crNetworkManager.ts
index fe1716c65..759a6c58d 100644
--- a/packages/playwright-core/src/server/chromium/crNetworkManager.ts
+++ b/packages/playwright-core/src/server/chromium/crNetworkManager.ts
@@ -348,6 +348,7 @@ export class CRNetworkManager {
     this._responseExtraInfoTracker.responseReceived(event);
 
     const request = this._requestIdToRequest.get(event.requestId);
+    console.log('eventLocation', this._serviceWorker ? 'sw' : 'page', 'fromServiceWorker:', event.response.fromServiceWorker, request?.request.url());
     // FileUpload sends a response without a matching request.
     if (!request)
       return;
```

We see the following with the test run as shown above:

```
[chromium] › library/chromium/chromium.spec.ts:42:6 › isServiceWorker() and serviceWorker() work
eventLocation page fromServiceWorker: false http://localhost:8907/serviceworkers/fetch/sw.html
eventLocation page fromServiceWorker: false http://localhost:8907/serviceworkers/fetch/style.css
eventLocation sw fromServiceWorker: false http://localhost:8907/serviceworkers/fetch/sw.js
eventLocation sw fromServiceWorker: false http://localhost:8907/request-from-within-worker.txt
eventLocation sw fromServiceWorker: false http://localhost:8907/inner.txt
eventLocation page fromServiceWorker: true http://localhost:8907/inner.txt
```

and the following when the context.route is added:

```
[chromium] › library/chromium/chromium.spec.ts:42:6 › isServiceWorker() and serviceWorker() work
eventLocation page fromServiceWorker: false http://localhost:8907/serviceworkers/fetch/sw.html
eventLocation page fromServiceWorker: false http://localhost:8907/serviceworkers/fetch/style.css
eventLocation sw fromServiceWorker: false http://localhost:8907/serviceworkers/fetch/sw.js
eventLocation sw fromServiceWorker: false http://localhost:8907/request-from-within-worker.txt
eventLocation sw fromServiceWorker: false http://localhost:8907/inner.txt
eventLocation page fromServiceWorker: true undefined
```

The issue likley starts earlier in `_onRequestWillBeSent`, but it highlights a larger fundamental issue to resolve around this feature: if a SW's `fetch` event handler handles a frame's request, how do we report the request(s). In the case above, the fetch requests seem "the same", but there's only one request that hits the network. Also, recall the Frame could call `fetch('/foo')`, and the SW could decide to:

1. serve from cache
2. run the fetch request as is (like in the example above)
3. rewrite it completely and hit the network as `/bar`, or multiple endpoints that it then merges into one Request, etc.

In our simple example, how many times do we expect to see `context.on('request', …)` fire for `/inner.txt`? I think the answer has to be 2 times since there is no guarantee frame and SW requests map 1:1 to eachother. Given this, we have to provide enough info on our Request object:

```
Request
   isServiceWorkerRequest(): the network was actually engaged within the SW.
   isFromServiceWorker(): this says nothing about if the network was hit or not. It just indicates the request came from the SW (i.e. via [`respondWith`](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/respondWith)).  
```

^ This is a bit of word soup for me to look at on Monday 😄

rwoll committed:
feat: add serviceWorkerPolicy option

Anecdotally, it's common for users to want to disable Service Workers
while testing, so this provides a simple context option to do so.

It also prepares us to land #14321 which requires us to have a point for
folks to enable seeing Service Worker Requests as events (and allow for
routing).

rwoll committed:
lint

rwoll committed:
add Response.isFromServiceWorker

rwoll committed:
Merge branch 'feat/swp' into feature/service-worker-support

rwoll commented:
Currently working on some of the opt-(in|out) logic. Once that's in, I'll update the description and open for review as that will be an actual candidate for landing and API review.

rwoll committed:
add serviceWorkerPolicy for opt-in SW networking

rwoll committed:
update docs

rwoll committed:
tests: make port agnostic, skip SW on older CR versions

rwoll committed:
lint

rwoll committed:
fix serviceWorkerPolicy

dgozman commented on packages/playwright-core/src/server/dispatchers/networkDispatchers.ts:
```diff
@@ -39,13 +40,15 @@ export class RequestDispatcher extends Dispatcher<Request, channels.RequestChann
       postData: postData === null ? undefined : postData.toString('base64'),
       headers: request.headers(),
       isNavigationRequest: request.isNavigationRequest(),
+      isServiceWorkerRequest: request.isServiceWorkerRequest(),
```
Why send this one over the protocol? It can be easily computed on the client side.

dgozman commented on packages/playwright-core/src/server/network.ts:
```diff
@@ -193,10 +199,18 @@ export class Request extends SdkObject {
+    return this._serviceWorker;
+  }
+
+  isServiceWorkerRequest(): boolean {
```
I'd remove this method, and instead have it on the client side.

dgozman commented on packages/playwright-core/src/server/types.ts:
```diff
@@ -373,3 +374,8 @@ export type AndroidDeviceOptions = {
   omitDriverInstall?: boolean,
 };
+
+export type ServiceWorkerPolicy = {
```
I'd suggest an enum with 3 values, something like default/disabled/inspected.

rwoll replied:
Updating—this was an artifact of an evolving policy where there were more options.

dgozman commented on packages/playwright-core/types/types.d.ts:
```diff
@@ -14616,6 +14649,14 @@ export interface Response {
+   * If set, indicates whether this Response was fullfilled by a Service Worker's Fetch Handler (i.e. via
+   * [FetchEvent.respondWith](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent/respondWith)).
+   */
+  isFromServiceWorker(): null|boolean;
```
Why `null` here? I'd return `false` for other browsers.

rwoll replied:
Unresolving—we're going to see if we can get the correct information here across all the browsers.

rwoll replied:
Per-offline comments, we'll land as it and follow up.

dgozman commented on docs/src/api/class-request.md:
```diff
@@ -85,6 +85,15 @@ Name of the header.
 
 Whether this request is driving frame's navigation.
 
+## method: Request.isServiceWorkerRequest
```
Are we sure we want this one? It is equivalent to `!!request.serviceWorker()` after all.

rwoll replied:
Agreed—removed.

dgozman commented on docs/src/api/class-request.md:
```diff
@@ -207,6 +216,15 @@ following: `document`, `stylesheet`, `image`, `media`, `font`, `script`, `texttr
+This field is Chromium only. It's safe to call when using other browsers, but it will always be `null`.
+:::
+
+The Service [`Worker`] that is performing the request.
```
```suggestion
The Service [Worker] that is performing the request.
```

For the proper link.

dgozman commented on docs/src/api/class-response.md:
```diff
@@ -51,6 +51,15 @@ Name of the header.
 
 Returns all values of the headers matching the name, for example `set-cookie`. The name is case insensitive.
 
+## method: Response.isFromServiceWorker
```
Bikeshedding: `fulfilledByServiceWorker`?

dgozman commented on docs/src/api/params.md:
```diff
@@ -636,10 +636,15 @@ contexts override the proxy, global proxy will be never used and can be any stri
+## context-option-service-worker-policy
+- `serviceWorkerPolicy` <[ServiceWorkerPolicy]>
+
+If set to `disabled`, all Service Worker registrations will be blocked.
```
This needs explanation for all other values as well.

dgozman commented on packages/playwright-core/src/client/network.ts:
```diff
@@ -225,6 +238,8 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
 
   private _raceWithPageClose(promise: Promise<any>): Promise<void> {
+    if (this.request().isServiceWorkerRequest())
+      return Promise.resolve();
```
I think we should race with sw close event here.

dgozman commented on packages/playwright-core/src/server/chromium/crBrowser.ts:
```diff
@@ -485,6 +470,8 @@ export class CRBrowserContext extends BrowserContext {
   async doUpdateRequestInterception(): Promise<void> {
     for (const page of this.pages())
       await (page._delegate as CRPage).updateRequestInterception();
+    for (const sw of this.serviceWorkers())
```
I think we should also cover `setExtraHTTPHeaders`, `setOffline` and `doSetHTTPCredentials`.

dgozman commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -241,7 +250,7 @@ export class CRNetworkManager {
     }
 
-    if (!frame) {
+    if (!frame && !this._serviceWorker) {
```
Let's add a comment that non-service-worker requests must have a frame. If we didn't find one, pretend there was not request :)

dgozman commented on packages/playwright-core/src/server/chromium/crServiceWorker.ts:
```diff
@@ -0,0 +1,80 @@
+    return this._networkManager.setRequestInterception(this._needsRequestInterception()).catch(e => { });
+  }
+
+  _needsRequestInterception(): boolean {
```
nit: I think this is called from CRNetworkManager, so remove underscore?

dgozman commented on utils/generate_types/overrides.d.ts:
```diff
@@ -336,5 +336,9 @@ export type AndroidKey =
 export const _electron: Electron;
 export const _android: Android;
 
+export type ServiceWorkerPolicy = 'default' | 'disabled' | {
```
We prefer not to export auxilary types unless we really need that.

rwoll committed:
remove isServiceWorker

rwoll committed:
s/isFromServiceWorker/fulfilledByServiceWorker

rwoll committed:
simplify and document serviceWorkerPolicy

rwoll committed:
fix lint error

rwoll committed:
s/_needsRequestInterception/needsRequestInterception

pavelfeldman commented on packages/playwright-core/src/server/browserContext.ts:
```diff
@@ -122,6 +122,20 @@ export abstract class BrowserContext extends SdkObject {
+
+        navigator.serviceWorker.register = () => {
+          console.warn('Service Worker registration disabled by Playwright');
+          return Promise.reject('Service Worker registration disabled by Playwright');
```
I would probably stall this one.

rwoll replied:
For stall here, do you mean return (A) a Promise that never resolves or (B) immediately return a resolved promise?

rwoll committed:
wip: setOffline, HTTP Creds, extra headers

rwoll committed:
fix HTTP Creds test

rwoll committed:
race with SW close

rwoll committed:
s/inspected/enable-network-events

rwoll committed:
fix rename

rwoll committed:
Merge branch 'main' into feature/service-worker-support

rwoll committed:
stall sw when policy is disabled

rwoll commented on tests/library/browsercontext-service-worker-policy.spec.ts:
```diff
@@ -0,0 +1,32 @@
+});
+
+it.describe('disabled', () => {
+  it.use({ serviceWorkerPolicy: 'disabled' });
```
s/serviceWorkerPolicy/serviceWorkers

rwoll committed:
make fulfilledByServiceWorker non-nullable (albeit broken on WK and FF…for now)

rwoll committed:
add fulfilledByServiceWorker test for all browsers, support WK

rwoll commented on docs/src/api/class-response.md:
```diff
@@ -22,6 +22,15 @@ Waits for this response to finish, returns always `null`.
+## method: Response.fulfilledByServiceWorker
+- returns: <[boolean]>
+
+:::note
+This field is Chromium only. It's safe to call when using other browsers, but it will always be `false` (for now).
+:::
```
```suggestion
```

rwoll committed:
remove invalid CORS test

Per https://www.w3.org/TR/service-workers/#origin-restriction and
observing browser behavior. We don't expect the browser to ever even
issue a request event if the main script is cross-origin.

rwoll commented on docs/src/api/params.md:
```diff
@@ -636,10 +636,19 @@ contexts override the proxy, global proxy will be never used and can be any stri
 on selectors that imply single target DOM element will throw when more than one element matches the selector.
 See [Locator] to learn more about the strict mode.
 
+## context-option-service-worker-policy
+- `serviceWorkerPolicy` <[ServiceWorkerPolicy]<"default"|"disabled"|"enable-network-events">>
+
+* `"default"`: Sites can register [Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API). However, the Network requests from within the Service Worker (as well as the main Service Worker script request itself) are not reported in [`method: BrowserContext.route`] nor [`event: BrowserContext.request`] event listeners.
+* `"disabled"`: Blocks all registration of Service Workers.
+* `"enable-network-events"`: Service Workers are enabled, and their Network Requests will show up in [`method: BrowserContext.route`] and the [`event: BrowserContext.request`] event listeners as well as the HAR.
+
+Defaults to `"default"`.
+
```
Per-offline discussion:

* change `serviceWorkersPolicy` to `serviceWorkers`
* options are `"allowed" | "blocked"`, defaults to `"allowed"`
* opt everyone in by default (based on some GH searches, we don't think the frame() throwing case will be hit)
* fix upstream bug with .response() hanging for errored SW Main Request scripts (e.g. 404, wrong content-type)

rwoll committed:
opt-in by default, s/serviceWorkersPolicy/serviceWorkers, reduce options

rwoll committed:
update docs

rwoll committed:
link upstream patch in fixme

rwoll committed:
Merge branch 'main' into feature/service-worker-support

rwoll committed:
add FF fulfilledByServiceWorker impl

rwoll committed:
fix FF expectation

rwoll commented:
Closing this to chop it up into a few different PRs
